path_login.go

```diff
@@ -267,6 +198,118 @@ func (b *jwtAuthBackend) pathLoginRenew(ctx context.Context, req *logical.Reques
 	return resp, nil
 }
 
+func (b *jwtAuthBackend) verifyToken(ctx context.Context, config *jwtConfig, role *jwtRole, rawToken string) (map[string]interface{}, error) {
+	allClaims := make(map[string]interface{})
+
+	provider, err := b.getProvider(ctx, config)
+	if err != nil {
+		return nil, errwrap.Wrapf("error getting provider for login operation: {{err}}", err)
+	}
+
+	oidcConfig := &oidc.Config{
+		SupportedSigningAlgs: config.JWTSupportedAlgs,
+	}
+
+	if role.RoleType == "oidc" {
+		oidcConfig.ClientID = config.OIDCClientID
+	} else {
+		oidcConfig.SkipClientIDCheck = true
+	}
+	verifier := provider.Verifier(oidcConfig)
+
+	idToken, err := verifier.Verify(ctx, rawToken)
+	if err != nil {
+		return nil, errwrap.Wrapf("error validating signature: {{err}}", err)
+	}
+
+	if err := idToken.Claims(&allClaims); err != nil {
+		return nil, errwrap.Wrapf("unable to successfully parse all claims from token: {{err}}", err)
+	}
+
+	if role.BoundSubject != "" && role.BoundSubject != idToken.Subject {
+		return nil, errors.New("sub claim does not match bound subject")
+	}
+	if len(role.BoundAudiences) > 0 {
+		var found bool
+		for _, v := range role.BoundAudiences {
+			if strutil.StrListContains(idToken.Audience, v) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, errors.New("aud claim does not match any bound audience")
+		}
+	}
+
+	if len(role.BoundClaims) > 0 {
+		for claim, expValue := range role.BoundClaims {
+			actValue := getClaim(allClaims, claim)
+			if actValue == nil {
+				return nil, fmt.Errorf("claim is missing: %s", claim)
+			}
+
+			if expValue != actValue {
+				return nil, fmt.Errorf("claim '%s' does not match associated bound claim", claim)
+			}
+		}
+	}
+
+	return allClaims, nil
+}
+
+// createIdentity creates an alias and set of groups aliass based on the role
+// definition and received claims.
+func (b *jwtAuthBackend) createIdentity(allClaims map[string]interface{}, role *jwtRole) (*logical.Alias, []*logical.Alias, error) {
+	userClaimRaw, ok := allClaims[role.UserClaim]
+	if !ok {
+		return nil, nil, fmt.Errorf("claim %q not found in token", role.UserClaim)
+	}
+	userName, ok := userClaimRaw.(string)
+	if !ok {
+		return nil, nil, fmt.Errorf("claim %q could not be converted to string", role.UserClaim)
+	}
+
+	metadata, err := extractMetadata(allClaims, role.ClaimMappings)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	alias := &logical.Alias{
+		Name:     userName,
+		Metadata: metadata,
+	}
+
+	var groupAliases []*logical.Alias
+
+	if role.GroupsClaim != "" {
```

Comment:
This is just a stylistic thing and feel free not to do it. I had an old coworker who was all about avoiding nesting if possible, because he thought nested logic could get confusing. In this situation, he would have encouraged me to simply do this here and return right away, merely so I could un-nest the remaining:
```
if role.GroupsClaim == "" { 
    return alias, groupAliases, nil
}
// Now everything else doesn't have to be nested.
```
Totally up to you. I don't really think it's super consequential or that the code is hard to reason about, but thought it might be interesting to bring up and see what you think, or if it even matters.

Reply:
I've done both ways, usually preferring not too much nested content. I think this code was nested before and I just didn't change it, but it would probably be better if restructured.